include/awkward/array/NumpyArray.h

```diff
@@ -63,6 +63,9 @@ namespace awkward {
 
     // operations
     const Index64 count64() const override;
+    // FIXME: merge to the above?
+    const Index64 count64(int64_t axis) const;
+
```

Comment:
@jpivarski - This is a temporary member function. The question is if the one above should take ```axis``` as a parameter.

Reply:
The purpose of `Index64 count64()` is internal: some operations can be defined more simply if they can see how many items per entry a node's `content_` has. It's a virtual method so that the node doesn't have to care what type of `content_` it has. Since the purpose of this is technical, it would only get called when you know that you're interested in exactly one level down. That's why it doesn't have an `axis` parameter.

It was originally introduced because I thought that the flattening of something like

```python
[[1.1, 2.2, 3.3], None, [4.4, 5.5], None, None, [6.6]]
```

should be

```python
[1.1, 2.2, 3.3, None, 4.4, 5.5, None, None, 6.6]
```

which would require knowledge of how many items each inner list has (3, 2, and then 1) to put in the appropriate gaps between the `None`s.

I would have needed this `count64` (without `axis`) to implement the above. However, the correct flattening of an OptionType(ListType) is

```python
[1.1, 2.2, 3.3, 4.4, 5.5, 6.6]
```

That is, a missing list should be treated like an empty list. (I have to remember now why I became so sure that this is the right answer after having implemented it the other way. It must be for consistency with something else—there are some type systems that even fail to distinguish between a missing value and an empty list...)

With this goal in mind, `count64` is not technically necessary, but I left it in because it's probable that some other method would need it. (I suspect that `getitem` in which the slices have ListType will need this method. Possibly also passing down `axis` parameters in other methods.)

Since we don't _need_ it yet, you can leave some implementations as `runtime_error("FIXME")`. In some cases, though, the implementation of `std::shared_ptr<Content> count(int64_t axis)` can be nicely factorized into a part that doesn't need to know `axis` and a part that does, and the transformation of an `Index64` (produced by the no-axis version) into a `NumpyArray` (required by the with-axis version) is a simple operation. (Unfortunately, though, type specialization of counts as 32-bit, unsigned 32-bit, and 64-bit sometimes eliminates the possibility of the one method simply calling the other.)